2.0-PROPOSAL/README.md

```diff
@@ -0,0 +1,885 @@
+| `PROJECT_USER` | Project | Read, create events for the specified project |
+| `PROJECT_DEVELOPER` | Project | Read and update the specified project |
+| `PROJECT_ADMIN` | Project | Read, modify, delete, and create events for the specified project; grant/revoke project-scoped roles to/from users and service accounts |
+
+N.B.: A principal in the global `PROJECT_CREATOR` automatically receives the
+`PROJECT_ADMIN` role for any new project they create.
+
+#### Secret Storage
+
+With respect to secrets used by Brigade itself, for instance bearer tokens
+associated with a user session or service account, it is both sufficient and
+easy to persist only a secure, one-way hash; however, no such assumption can be
+made with respect to a _project's_ secrets. Without knowing _how_ a project will
+utilize its secrets, any encryption must be reversible such that the original,
+cleartext value is available to workers and jobs in the course of
+event-handling. This being the case, secure storage of encrypted project-level
+secrets is a necessity. Implementing this would prove onerous once factors such
+as periodic key rotation have been accounted for and the author does not propose
+such an undertaking. On the surface, delegating secret storage to a service such
+as Azure Key Vault or HashiCorp Vault may appear easy and attractive, but would
+further increase the operational footprint of Brigade's infrastructure
+dependencies and impede ease of use for evaluation or development purposes,
+making this another undesirable option.
+
+Recognizing, however, that once created, project-level secrets are utilized
+exclusively by workers and jobs, both of which execute on the underlying
+workload execution substrate (i.e. Kubernetes), and further recognizing that to
+be made available to those workers and pods, secret values must necessarily, at
+some point, be stored using the substrate's native secret storage mechanisms
+(i.e. Kubernetes `Secret` resources), it begins to appear tenable to delegate
+secret storage _exclusively_ to the substrate via the scheduling subsystem. i.e.
+When setting the value of a secret, it is set directly on the substrate and
+stored nowhere else.
+
+The secret storage solution proposed above means that Brigade can do _no worse_
+than the underlying substrate in securing a project's secrets. If the underlying
+substrate's (i.e. Kubernetes') native secret storage facilities are deemed
+inadequate, that problem is not Brigade's and can be addressed at a lower level,
+since Kubernetes supports pluggable backends for secret management.
+
+#### Preventing Escalation of Privileges
+
+Isolating each Brigade project to its own namespace has been among Brigade's
+most requested features. To fulfill this, Brigade 2.0 is proposed to (via the
+scheduling subsystem) _automatically create a brand new Kubernetes namespace for
+each new project_.
```

Comment:
From @brandonkal:

> Automatic namespace creation will make Brigade unusable in certain environments such as [https://okteto.com/](https://okteto.com/) and similar services.

Earlier sections of the proposal emphasize that Brigade could be offering a lot of value to people who lack Kubernetes expertise if the de facto requirements for said experience and direct access to a cluster were dropped. To whatever extent is achievable, Kubernetes is meant to fade into the background as an _implementation detail_. You can then view Brigade 2.0 as something much closer to a PaaS than 1.x was.

Viewed in this context, I think Brigade 2.0 and a service like Okteto are mutually exclusive.

__Okteto:__ Small k8s sandboxes (one namespace) aaS for developers who know Kubernetes.

__Brigade 2.0:__ Small event-driven scripting sandboxes (one project) aaS for developers _in general_. k8s is an implementation detail.

> Brigade should not rely on having the permissions to create and manage namespaces.

Perhaps the proposal needs to be more explicit about this, but there was a series of cascading design decisions, each necessitating the next, with this being just about the last in the chain. If Brigade cannot create new namespaces, pretty much the whole proposal falls apart.

Here is the rationale:

1. Today Brigade 1.x is subject to data loss and resource contention issues because of Kubernetes being used as a makeshift document store and message bus.
1. Introducing a real document store and message bus solves those issues.
1. This means clients like the `brig` CLI and gateways can no longer communicate directly with the k8s API server.
1. This means Brigade 2.0 needs the full-featured API that Brigade 1.x lacked.
1. If we're going to have an API, there needs to be a way to authenticate users and authorize actions.
1. But there is no way for random APIs to delegate authn or authz to Kubernetes.
    1. We spent a lot of time trying to figure out how this could be done and ultimately concluded that it can not be done currently. In fact, we discovered along the way that it was proposed feature of Kubernetes at one time, but appears to have long since been abandoned.
    1. Even if this were possible, it undermines our goal of making Brigade 2.0 less k8s-centric and _not_ requiring users to have direct access to a cluster.
1. Authn and authz would have to be achieved through other means.
1. The Brigade 2.0 API doesn't know what permissions (if any) a user has within the k8s cluster.
1. Allowing a user to "bring their own namespace" by telling Brigade what namespace to use for their project would create the possibility of completely hijacking someone else's namespace.
1. The solution is that Brigade 2.0 creates _fresh_ namespaces for each project. It's proposed that it can at least tell you the name of that namespace so that _if_ you have sufficient permissions within the cluster, you can go tweak the namespace, the service accounts therein, etc. to your heart's content.

Hopefully, you can see how _not_ permitting Brigade to create namespaces unwinds this entire stack and would put us back at square one.

An alternative that I will dismiss now is having everything happen in just one namespace as in Brigade 1.x (which incidentally was not meant to work that way-- it's actually a bug). _Many_ users have asked to have projects isolated to their own namespaces. Delivering on that is a high priority.

Reply:
It seems to me this could easily be solved by creating a Project CRD. It wouldn't actually have to store the Project config (just the name). Then if a user has permissions to create that resource in their chosen namespace, brigade would launch pods for that project etc in that namespace. It's odd that the existing design has namespace specified anywhere but metadata.namespace

Reply:
> It seems to me this could easily be solved by creating a Project CRD. Then if a user has permissions to create that resource in their chosen namespace, brigade would launch pods for that project etc in that namespace.

I will re-iterate that there is a goal for this to be useful to people who don't have k8s experience or (direct) cluster access.